examples/node/store.c

```diff
@@ -0,0 +1,226 @@
+/* Copyright (c) 2019, Codership Oy. All rights reserved.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 2 of the License.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
+ */
+
+#include "store.h"
+
+#include "log.h"
+
+#include <pthread.h>
+#include <stdlib.h>   // abort()
+#include <string.h>   // memset()
+#include <stdint.h>   // uintptr_t
+#include <errno.h>
+#include <assert.h>
+
+struct node_store
+{
+    wsrep_gtid_t    gtid;
+    pthread_mutex_t gtid_mtx;
+    wsrep_trx_id_t  trx_id;
+    pthread_mutex_t trx_id_mtx;
+};
+
+// don't bother with dynamic allocation with a single store
+static struct node_store _store =
+{
+    .gtid       = {{{0,}}, WSREP_SEQNO_UNDEFINED} /*WSREP_GTID_UNDEFINED*/,
+    .gtid_mtx   = PTHREAD_MUTEX_INITIALIZER,
+    .trx_id     = 0,
+    .trx_id_mtx = PTHREAD_MUTEX_INITIALIZER
+};
+
+static struct node_store* _store_ptr = &_store;
+
+// stub
+node_store_t*
+node_store_open(const struct node_options* const opts)
+{
+    (void)opts;
+    node_store_t* ret = _store_ptr;
+    _store_ptr = NULL;
+    return ret;
+}
+
+// stub
+void
+node_store_close(struct node_store* store)
+{
+    assert(store);
+    (void)store;
+}
+
+int
+node_store_init(struct node_store*  const store,
+                const wsrep_gtid_t* const gtid)
+{
+    assert(store);
+
+    int ret;
+    ret = pthread_mutex_lock(&store->gtid_mtx);
+    if (ret)
+    {
+        NODE_FATAL("Failed to lock store GTID");
+        abort();
+    }
+
+    store->gtid = *gtid;
+    pthread_mutex_unlock(&store->gtid_mtx);
+
+    return ret;
+}
+
+void
+node_store_gtid(struct node_store* const store,
+                wsrep_gtid_t*      const gtid)
+{
+    assert(store);
+
+    int ret;
+    ret = pthread_mutex_lock(&store->gtid_mtx);
+    if (ret)
+    {
+        NODE_FATAL("Failed to lock store GTID");
+        abort();
+    }
+
+    *gtid = store->gtid;
+    pthread_mutex_unlock(&store->gtid_mtx);
+}
+
+// stub
+int
+node_store_prepare_execute(node_store_t*   const store,
```

Comment:
I would rename this to `node_store_start_execute()` or `node_store_begin_execute()` to avoid using "prepare" which could be confused to 2PC prepare.

Reply:
I don't agree with the notion that 2PC usurped a common English word "prepare", but "start" and "begin" are just plain wrong. So nothing there!